src/models/Ruleserver.ts

```diff
@@ -0,0 +1,290 @@
+            // the events in the previousState and event slots will be distinct events.
+            // In either case we can therefore use previousState to get the right event id to stop.
+            const entry: EventRules|undefined = this.getEventRules(change.event.room_id, change.previousState.event_id);
+            if (entry === undefined) {
+                LogService.error('RuleServer', `Could not find the rules for the previous modified state ${change.event['state_type']} ${change.event['state_key']} ${change.previousState?.event_id}`);
+                return;
+            }
+            this.stopEventRules(entry);
+        }
+    }
+
+    /**
+     * Watch the ban list for changes and serve its policies as rules.
+     * @param banList a BanList to watch for rule changes with.
+     */
+    public watch(banList: BanList): void {
+        banList.on('update', this.banListUpdateListener);
+    }
+
+    /**
+     * Remove all of the rules that have been created from the policies in this banList.
+     * @param banList The BanList to unwatch.
+     */
+    public unwatch(banList: BanList): void {
+        banList.removeListener('update', this.banListUpdateListener);
+        const listRules = this.rulesByEvent.get(banList.roomId);
+        this.nextToken();
+        if (listRules) {
+            [...listRules.values()].forEach(this.stopEventRules, this);
+        }
+    }
+
+    /**
+     * Process the changes that have been made to a BanList, this will ususally be called as a callback from `BanList.onChange`.
+     * @param banList The BanList that the changes happened in.
+     * @param changes An array of ListRuleChanges.
+     */
+    private update(banList: BanList, changes: ListRuleChange[]) {
+        if (changes.length > 0) {
+            this.nextToken();
+            changes.forEach(this.applyRuleChange, this);
+        }
+    }
+
+    /**
+     * Get all of the new rules since the token.
+     * @param sinceToken A token that has previously been issued by this server.
+     * @returns An object with the rules that have been started and stopped since the token and a new token to poll for more rules with.
+     */
+    public getUpdates(sinceToken: string | null): {start: RuleserverRule[], stop: string[], reset?: boolean, since: string} {
+        const updatesSince = <T = EventRules|string>(token: number | null, policyStore: T[][]): T[] => {
+            if (token === null) {
+                return policyStore.flat();
+            } else {
+                if (token === this.currentToken) {
+                    return [];
+                } else {
+                    return policyStore.slice(token).flat();
+                }
+            }
+        }
+        const [serverId, since] = sinceToken ? sinceToken.split('::') : [null, null];
+        const parsedSince: number | null = since ?  parseInt(since, 10) : null;
+        if (serverId && serverId !== this.serverId) {
```

Comment:
Shouldn't you also reset if `serverId == null`?

Reply:
Good question, I don't know. I was assuming that the client wouldn't have ever communicated with this server before (in this life or another) in this situation, but there might be an argument for being safer than sorry.